File: app/src/main/kotlin/br/com/victorwads/goldenunicorn/data/repositories/BanksRepository.kt
package br.com.victorwads.goldenunicorn.data.repositories

import android.content.Context
import br.com.victorwads.goldenunicorn.data.firebase.Collections
import br.com.victorwads.goldenunicorn.data.models.Bank
import br.com.victorwads.goldenunicorn.prepareCompare
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Source
import com.google.firebase.firestore.toObject
import java.util.Date
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine
class BanksRepository(context: Context) {

    private val bd = FirebaseFirestore.getInstance()
    private val ref = bd.collection(Collections.Banks)
    private val localStorage = context
        .getSharedPreferences(Collections.FirestoreCacheKeys, Context.MODE_PRIVATE)

    private fun shouldUseCache() = localStorage.getLong(lastUpdateKey, 0).let {
        (Date().time - it) < cacheDuration
    }

    private fun setLastUpdate() = with(localStorage.edit()) {
        putLong(lastUpdateKey, Date().time)
        apply()
    }

    suspend fun getAll(forceCache: Boolean = false) = suspendCoroutine<List<Bank>> { thread ->
        val source = if(forceCache || shouldUseCache()) Source.CACHE
        else Source.DEFAULT
        if(source == Source.DEFAULT) setLastUpdate()
        ref.get(source).addOnCompleteListener {task ->
            if(task.isSuccessful){
                thread.resume(task.result.documents.mapNotNull {
                    it.toObject<Bank>()
                }.sortedBy { it.name })
            } else {
                thread.resume(arrayListOf())
            }
        }
    }

    suspend fun getFiltered(search: String): List<Bank> {
        val normalizedSearch = search.prepareCompare()
        val allBanks = getAll(true)
        return allBanks.filter { bank ->
            bank.name.prepareCompare().contains(normalizedSearch) ||
            bank.fullName.prepareCompare().contains(normalizedSearch)
        }
    }

    companion object {
        const val lastUpdateKey = "lastBanksUpdate"
        const val cacheDuration = 2592000000// 30 * 24 * 60 * 60 * 1000;
    }
}

File: Sources/Data/repositories/BanksRepository.swift
import Foundation
import FirebaseFirestore
import FirebaseCrashlytics

class BanksRepository {

    static private let lastUpdateKey = "lastBanksUpdate"
    static private let cacheDuration: CGFloat = 2592000000

    private let db: Firestore
    private let collectionRef: CollectionReference
    private let userDefaults = UserDefaults.standard

    init() {
        db = Firestore.firestore()
        collectionRef = db.collection(Collections.Banks)
    }

    private func shouldUseCache() -> Bool {
        if let lastUpdate = userDefaults.value(forKey: BanksRepository.lastUpdateKey) as? CGFloat {
            return (Date().timeIntervalSince1970 - lastUpdate) < BanksRepository.cacheDuration
        }
        return false
    }

    private func setLastUpdate() {
        userDefaults.set(CGFloat(Date().timeIntervalSince1970), forKey: BanksRepository.lastUpdateKey)
    }

    func getAll(forceCache: Bool = false, _ completion: @escaping ([Bank]) -> Void) {
        let source: FirestoreSource = forceCache || shouldUseCache() ? .cache : .default
        if source == .default {
            setLastUpdate()
        }
        collectionRef.getDocuments(source: source) { (snapshot, error) in
            if let error = error {
                Crashlytics.crashlytics().record(error: error)
                completion([])
            } else {
                var banks: [Bank] = []
                for document in snapshot!.documents {
                    if let bank = try? document.data(as: Bank.self) {
                        banks.append(bank)
                    }
                }
                banks.sort { $0.name < $1.name }
                completion(banks)
            }
        }
    }

    func getFiltered(search: String, completion: @escaping ([Bank]) -> Void) {
        let normalizedSearch = search.prepareCompare()
        getAll(forceCache: true) { (banks) in
            if(search.isEmpty) {
                completion(banks)
                return
            }
            let filteredBanks = banks.filter { (bank) in
                return bank.name.prepareCompare().contains(normalizedSearch) ||
                bank.fullName.prepareCompare().contains(normalizedSearch)
            }
            completion(filteredBanks)
        }
    }
}

